rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::collections::VecDeque;
+use std::fmt;
+use std::time::{Duration, SystemTime};
+
+#[derive(Debug, Clone, Copy)]
+pub(super) struct Throughput {
+    bytes_read: f64,
+    per_time_elapsed: Duration,
+}
+
+impl Throughput {
+    pub(super) fn new(bytes_read: f64, per_time_elapsed: Duration) -> Self {
+        debug_assert!(
+            !bytes_read.is_nan(),
+            "cannot create a throughput if bytes_read == NaN"
+        );
+        debug_assert!(
+            bytes_read.is_finite(),
+            "cannot create a throughput if bytes_read == Inf"
+        );
+        debug_assert!(
+            !per_time_elapsed.is_zero(),
+            "cannot create a throughput if per_time_elapsed == 0"
+        );
+
+        Self {
+            bytes_read,
+            per_time_elapsed,
+        }
+    }
+
+    pub(super) fn per_time_elapsed(&self) -> Duration {
+        self.per_time_elapsed
+    }
+
+    pub(super) fn bytes_per_second(&self) -> f64 {
+        let per_time_elapsed_secs = self.per_time_elapsed.as_secs_f64();
+        if per_time_elapsed_secs == 0.0 {
+            return 0.0; // Avoid dividing by zero.
+        };
+
+        self.bytes_read / per_time_elapsed_secs
+    }
+
+    fn normalize_to_per_one_second(self) -> Self {
+        if self.per_time_elapsed == Duration::from_secs(1) {
+            // Already normalized to per second, return self without modifying it.
+            self
+        } else {
+            // Normalize to per second and return the result.
+            Self {
+                bytes_read: self.bytes_per_second(),
+                per_time_elapsed: Duration::from_secs(1),
+            }
+        }
+    }
+}
+
+impl PartialEq for Throughput {
+    fn eq(&self, other: &Self) -> bool {
+        if self.per_time_elapsed == other.per_time_elapsed {
+            // If both are using the same unit of time, compare bytes per second.
+            self.bytes_per_second() == other.bytes_per_second()
+        } else {
+            // Otherwise, normalize the unit of time and then compare adjusted bytes per second.
+            let normalized_self = self.normalize_to_per_one_second();
+            let normalized_other = other.normalize_to_per_one_second();
+            normalized_self.bytes_per_second() == normalized_other.bytes_per_second()
```

Comment:
I think we can compare `bytes_per_second` in all cases right? That's already been normalized and compares `per_time_elapsed`